observability/src/main/java/io/grpc/observability/interceptors/GcpLogHelper.java

```diff
@@ -0,0 +1,430 @@
+
+    /**
+     * Logs cancellation.
+     */
+    void logCancel(
+        long seqId,
+        String serviceName,
+        String methodName,
+        GrpcLogRecord.EventLogger eventLogger,
+        String rpcId) {
+      checkNotNull(serviceName, "serviceName");
+      checkNotNull(methodName, "methodName");
+      checkNotNull(rpcId, "rpcId");
+
+      GrpcLogRecord.Builder logEntryBuilder = createTimestamp()
+          .setSequenceId(seqId)
+          .setServiceName(serviceName)
+          .setMethodName(methodName)
+          .setEventType(EventType.GRPC_CALL_CANCEL)
+          .setEventLogger(eventLogger)
+          .setLogLevel(LogLevel.LOG_LEVEL_DEBUG)
+          .setRpcId(rpcId);
+      logger.log(Level.FINE, "Writing Cancel event to GcpLogSink");
+      sink.write(logEntryBuilder.build());
+    }
+  }
+
+  // Copied from internal
+  static final class ByteArrayMarshaller implements Marshaller<byte[]> {
+    @Override
+    public InputStream stream(byte[] value) {
+      return new ByteArrayInputStream(value);
+    }
+
+    @Override
+    public byte[] parse(InputStream stream) {
+      try {
+        return parseHelper(stream);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    private byte[] parseHelper(InputStream stream) throws IOException {
+      try {
+        return IoUtils.toByteArray(stream);
+      } finally {
+        stream.close();
+      }
+    }
+  }
+
+  // Copied from internal
+  static final class IoUtils {
+    /** maximum buffer to be read is 16 KB. */
+    private static final int MAX_BUFFER_LENGTH = 16384;
+
+    /** Returns the byte array. */
+    public static byte[] toByteArray(InputStream in) throws IOException {
+      ByteArrayOutputStream out = new ByteArrayOutputStream();
+      copy(in, out);
+      return out.toByteArray();
+    }
+
+    /** Copies the data from input stream to output stream. */
+    public static long copy(InputStream from, OutputStream to) throws IOException {
+      // Copied from guava com.google.common.io.ByteStreams because its API is unstable (beta)
+      checkNotNull(from);
+      checkNotNull(to);
+      byte[] buf = new byte[MAX_BUFFER_LENGTH];
+      long total = 0;
+      while (true) {
+        int r = from.read(buf);
+        if (r == -1) {
+          break;
+        }
+        to.write(buf, 0, r);
+        total += r;
+      }
+      return total;
+    }
+  }
+
+
+  static final class PayloadBuilder<T> {
+    T proto;
+    int size;
+
+    private PayloadBuilder(T proto, int size) {
+      this.proto = proto;
+      this.size = size;
+    }
+  }
+
+  static PayloadBuilder<GrpcLogRecord.Metadata.Builder> createMetadataProto(Metadata metadata) {
+    checkNotNull(metadata, "metadata");
+    GrpcLogRecord.Metadata.Builder metadataBuilder = GrpcLogRecord.Metadata.newBuilder();
+    byte[][] serialized = InternalMetadata.serialize(metadata);
+    int bytesAfterAdd = 0;
+    if (serialized != null) {
+      int curBytes = 0;
```

Comment:
`curBytes` is set to 0 and never updated later! I wonder what this code is doing and I suspect there is a bug. I suggest:

- explain what `bytesAfterAdd` is for - either name the var properly (is "add" for address, add, ...?) or add a comment what it is
- if curBytes is always 0 just remove it or figure out where is it supposed to be modified
- create a unit test to test this complex logic - and use the unit test to fix the bug

Reply:
Thanks for catching the bug. 
Updated the method with comments and test [metadataToProto](https://github.com/grpc/grpc-java/blob/38ca751347f3520032db07d134b690c6e84cfdea/observability/src/test/java/io/grpc/observability/interceptors/LogHelperTest.java#L177) is used to check if the logic of calculating size of Metadata. `nonEmptyMetadataSize` is hardcoded as calculating size of Metadata is dependent on the implementation.